### PR TITLE
feat: Adding options to deploy Promitor with managed identity

### PR DESCRIPTION
**Pull request by Mimetis**
# Adding options to authenticate Promitor using Managed Identity

> **[Update 1/4/2021]: Update yaml examples to the last PR commit**

Update of the charts (**Scraper** and **Resource-Discovery**) to allow user to authenticate **Promitor** with **Managed Identity**.

Basically, in both charts, `values.yaml`, `configmap.yaml`, `deployment.yaml`, and `secret.yaml` have been updated.

From a user perspective, here is a sample of a `values.yaml` that can be submitted:

Using a **Service Principal**:
``` yaml
azureAuthentication:
  mode: "ServicePrincipal"
  identity:
     id: "xxxxxxxx-xxxxx-xxxxxx-xxxxxx"
     key: "xxxxxxxx-xxxxx-xxxxxx-xxxxxx"
```

Using a **Service Managed Identity** (with aad pod indentity already deployed in the aks cluster):
``` yaml
azureAuthentication:
  mode: "SystemAssignedManagedIdentity"
  identity:
     binding: "aadpodid1"
```

Using a **User Managed Identity**:
``` yaml
azureAuthentication:
  mode: "UserAssignedManagedIdentity"
  identity:
     id: "xxxxxxxx-xxxxx-xxxxxx-xxxxxx"
     key: "xxxxxxxx-xxxxx-xxxxxx-xxxxxx"
     binding: "aadpodid1"
```

*Note* : 
- If `mode` is not specified, the `servicePrincipal` value is used by default.
- If `identity.key` is omitted (and needed when mode is `UserAssignedManagedIdentity` or `ServicePrincipal`), the environment variable `PROMITOR_AUTH_APPKEY` is used

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Thanks for the PR! Let's put this PR on hold until https://github.com/tomkerkhove/promitor/pull/1544 is finished

File: promitor-agent-resource-discovery/values.yaml [hunk 1]
```diff
@@ -12,8 +12,11 @@ image:
   tag:
 
 azureAuthentication:
-  appId:
-  appKey:
+  appId: ""
+  appKey: ""
+  managedIdentityId: ""
+  mode: "servicePrincipal"
+  aadpodidbinding: ""
```
Comment (tomkerkhove) [inline promitor-agent-resource-discovery/values.yaml:19]: I'd love to use this approach instead:
```suggestion
  mode: "ServicePrincipal"
  identity
    id: ""
    key: ""
    binding
```

**Review by tomkerkhove — COMMENTED** [review]
Are you able to update this to the recent changes or do you prefer if I pick this up?

**Comment by Mimetis** [comment]
No worry, I'll do it !

File: promitor-agent-resource-discovery/values.yaml (hunk 1 above)
Comment (Mimetis) [reply promitor-agent-resource-discovery/values.yaml:19]: Sounds good to me

File: promitor-agent-resource-discovery/values.yaml [hunk 2]
```diff
@@ -12,8 +12,11 @@ image:
   tag:
 
 azureAuthentication:
-  appId:
-  appKey:
```
Comment (tomkerkhove) [inline promitor-agent-resource-discovery/values.yaml:16]: We'll have to be backward compatible here where the new one is favored over the old one; but we shouldn't break the current experience.

File: promitor-agent-scraper/values.yaml [hunk 3]
```diff
@@ -13,8 +13,11 @@ image:
   tag:
 
 azureAuthentication:
-  appId: ""
-  appKey: ""
```
Comment (tomkerkhove) [inline promitor-agent-scraper/values.yaml:17]: We'll have to be backward compatible here where the new one is favored over the old one; but we shouldn't break the current experience.

File: promitor-agent-resource-discovery/README.md [hunk 4]
```diff
@@ -63,8 +63,10 @@ their default values.
 | `azureLandscape.tenantId`  | Id of Azure tenant to discover resources in |             |
 | `azureLandscape.subscriptions`  | List of Azure subscription ids to discover resources in | `[]`            |
 | `resourceDiscoveryGroups`  | List of resource discovery groups to configured following the [resource discovery declaration docs](https://promitor.io/configuration/v2.x/resource-discovery) |        |
-| `azureAuthentication.appId`  | Id of the Azure AD entity to authenticate with |             |
-| `azureAuthentication.appKey`  | Secret of the Azure AD entity to authenticate with |             |
```
Comment (tomkerkhove) [inline promitor-agent-resource-discovery/README.md:67]: We'll have to be backward compatible here where the new one is favored over the old one; but we shouldn't break the current experience.

We can thus keep this but mark it as deprecated with reference to new appraoch

File: promitor-agent-scraper/README.md [hunk 5]
```diff
@@ -60,8 +60,10 @@ their default values.
 | `image.tag`  | Tag of image to use | None, chart app version is used by default            |
 | `image.pullPolicy`  | Policy to pull image | `Always`            |
 | `image.pullSecrets`  | ImagePullSecrets for the pod | `[]`            |
-| `azureAuthentication.appId`  | Id of the Azure AD entity to authenticate with |             |
-| `azureAuthentication.appKey`  | Secret of the Azure AD entity to authenticate with |             |
```
Comment (tomkerkhove) [inline promitor-agent-scraper/README.md:64]: We'll have to be backward compatible here where the new one is favored over the old one; but we shouldn't break the current experience.

We can thus keep this but mark it as deprecated with reference to new appraoch

File: promitor-agent-scraper/README.md [hunk 6]
```diff
@@ -140,8 +141,8 @@ Specify each parameter using the `--set key=value[,key=value]` argument to
 
 ```console
 $ helm install promitor-agent-scraper promitor/promitor-agent-scraper \
-               --set azureAuthentication.appId='<azure-ad-app-id>' \
-               --set azureAuthentication.appKey='<azure-ad-app-key>' \
+               --set azureAuthentication.identity.id='<azure-ad-app-id>' \
```
Comment (tomkerkhove) [inline promitor-agent-scraper/README.md:144]: Just to be complete
```suggestion
               --set azureAuthentication.identity.mode='ServicePrincipal' \
               --set azureAuthentication.identity.id='<azure-ad-app-id>' \
```

File: promitor-agent-resource-discovery/README.md [hunk 7]
```diff
@@ -30,8 +30,8 @@ To install the chart with the release name `promitor-agent-resource-discovery`:
 
 ```console
 $ helm install promitor-agent-resource-discovery promitor/promitor-agent-resource-discovery \
-               --set azureAuthentication.appId='<azure-ad-app-id>' \
-               --set azureAuthentication.appKey='<azure-ad-app-key>' \
+               --set azureAuthentication.identity.id='<azure-ad-app-id>' \
```
Comment (tomkerkhove) [inline promitor-agent-resource-discovery/README.md:33]: Just to be complete
```suggestion
               --set azureAuthentication.identity.mode='ServicePrincipal' \
               --set azureAuthentication.identity.id='<azure-ad-app-id>' \
```

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
LGTM, just a few small remarks to be backwards compatible

File: promitor-agent-resource-discovery/values.yaml [hunk 8]
```diff
@@ -12,8 +12,13 @@ image:
...
+  appId: "" # [Deprecated] Prefer identity.id
+  appKey: "" # [Deprecated] Prefer identity.key
+  mode: "servicePrincipal"
```
Comment (tomkerkhove) [inline promitor-agent-resource-discovery/values.yaml:17]:
```suggestion
  mode: "ServicePrincipal"
```

File: promitor-agent-scraper/templates/configmap.yaml [hunk 9]
```diff
@@ -11,6 +11,13 @@ data:
   runtime.yaml: |-
       server:
         httpPort: {{ .Values.service.targetPort | quote }}
+      authentication:
+        mode: {{ .Values.azureAuthentication.mode | default "servicePrincipal"}}
```
Comment (tomkerkhove) [inline promitor-agent-scraper/templates/configmap.yaml:15]:
```suggestion
        mode: {{ .Values.azureAuthentication.mode | default "ServicePrincipal"}}
```

File: promitor-agent-resource-discovery/templates/configmap.yaml [hunk 10]
```diff
@@ -12,6 +12,13 @@ data:
   runtime.yaml: |-
       server:
         httpPort: {{ .Values.service.targetPort | quote }}
+      authentication:
+        mode: {{ .Values.azureAuthentication.mode | default "servicePrincipal"}}
```
Comment (tomkerkhove) [inline promitor-agent-resource-discovery/templates/configmap.yaml:16]:
```suggestion
        mode: {{ .Values.azureAuthentication.mode | default "ServicePrincipal"}}
```

File: promitor-agent-scraper/values.yaml [hunk 11]
```diff
@@ -13,8 +13,13 @@ image:
...
+  appId: "" # [Deprecated] Prefer identity.id
+  appKey: "" # [Deprecated] Prefer identity.key
+  mode: "servicePrincipal"
```
Comment (tomkerkhove) [inline promitor-agent-scraper/values.yaml:18]:
```suggestion
  mode: "ServicePrincipal"
```

**Review by tomkerkhove — APPROVED** [review]
LGTM, thank you a lot @Mimetis!

**Comment by locmai** [comment]
Hi @Mimetis @tomkerkhove  , I have tried resource discovery chart and got authenticated with aadpodbinding for a Us Gov cloud subscription: 
Screen Shot 2021-05-26 at 09 46 37

But when I tried the health check api with the `api/v1/health?includeDependencies=true` and got this:
```
Unable to discover resources due to an unexpected exception: Operation returned an invalid status code 'Unauthorized'
Microsoft.Azure.Management.ResourceGraph.Models.ErrorResponseException: Operation returned an invalid status code 'Unauthorized'
at Microsoft.Azure.Management.ResourceGraph.ResourceGraphClient.ResourcesWithHttpMessagesAsync(QueryRequest query, Dictionary`2 customHeaders, CancellationToken cancellationToken)
at Microsoft.Azure.Management.ResourceGraph.ResourceGraphClientExtensions.ResourcesAsync(IResourceGraphClient operations, QueryRequest query, CancellationToken cancellationToken)
at Promitor.Agents.ResourceDiscovery.Graph.AzureResourceGraph.<>c__DisplayClass13_0.<<QueryAsync>b__0>d.MoveNext() in /src/Promitor.Agents.ResourceDiscovery/AzureResourceGraph.cs:line 80
--- End of stack trace from previous location where exception was thrown ---
at Promitor.Agents.ResourceDiscovery.Graph.AzureResourceGraph.<>c__DisplayClass14_0`1.<<InteractWithAzureResourceGraphAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at Polly.Retry.AsyncRetryEngine.ImplementationAsync[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Func`5 onRetryAsync, Int32 permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider, Boolean continueOnCapturedContext)
at Polly.AsyncPolicy.ExecuteAsync[TResult](Func`3 action, Context context, CancellationToken cancellationToken, Boolean continueOnCapturedContext)
at Promitor.Agents.ResourceDiscovery.Graph.AzureResourceGraph.InteractWithAzureResourceGraphAsync[TResponse](String queryName, String query, List`1 targetSubscriptions, Func`2 interactionFunc) in /src/Promitor.Agents.ResourceDiscovery/AzureResourceGraph.cs:line 93
at Promitor.Agents.ResourceDiscovery.Graph.AzureResourceGraph.QueryAsync(String queryName, String query, List`1 targetSubscriptions) in /src/Promitor.Agents.ResourceDiscovery/AzureResourceGraph.cs:line 77
at Promitor.Agents.ResourceDiscovery.Graph.AzureResourceGraph.QueryForResourcesAsync(String queryName, String query, List`1 targetSubscriptions) in /src/Promitor.Agents.ResourceDiscovery/AzureResourceGraph.cs:line 67
at Promitor.Agents.ResourceDiscovery.Health.AzureResourceGraphHealthCheck.ProbeSubscriptionAsync(String query, String subscription) in /src/Promitor.Agents.ResourceDiscovery/AzureResourceGraphHealthCheck.cs:line 84
```

Is there any known issue with Us Gov cloud running with aadpodidentity? 

P/S: I have checked the identity (with Reader and even Contributor role for the subscription) with azure-cli and I could fetch everything.

**Comment by tomkerkhove** [comment]
Thanks for letting me know, would you mind opening a discussion for this please? I'll then take a look!

**Pull request merged**